tests/study_tests/test_multi_objective.py

```diff
@@ -0,0 +1,105 @@
+from typing import List
+from typing import Sequence
+
+import pytest
+
+import optuna
+from optuna.study import StudyDirection
+from optuna.study._multi_objective import _dominates
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+def _create_trial(values: Sequence[float], state: TrialState = TrialState.COMPLETE) -> FrozenTrial:
+    return optuna.trial.create_trial(values=list(values), state=state)
+
+
+@pytest.mark.parametrize(
+    ("v1", "v2"), [(-1, 1), (-float("inf"), 0), (0, float("inf")), (-float("inf"), float("inf"))]
+)
+def test_dominates_1d_not_equal(v1: float, v2: float) -> None:
+    t1 = _create_trial([v1])
+    t2 = _create_trial([v2])
+
+    assert _dominates(t1, t2, [StudyDirection.MINIMIZE])
+    assert not _dominates(t2, t1, [StudyDirection.MINIMIZE])
+
+    assert _dominates(t2, t1, [StudyDirection.MAXIMIZE])
+    assert not _dominates(t1, t2, [StudyDirection.MAXIMIZE])
+
+
+@pytest.mark.parametrize("v", [0, -float("inf"), float("inf")])
+@pytest.mark.parametrize("dir", [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE])
+def test_dominates_1d_equal(v: float, dir: StudyDirection) -> None:
+    assert not _dominates(_create_trial([v]), _create_trial([v]), [dir])
+
+
+def test_dominates_2d() -> None:
+    directions = [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE]
+
+    # Check all pairs of trials consisting of these values, i.e.,
+    # [-inf, -inf], [-inf, -1], [-inf, 1], [-inf, inf], [-1, -inf], ...
+    # These values should be specified in ascending order.
+    vals = [-float("inf"), -1, 1, float("inf")]
+
+    # This function converts integer indices to values.
+    def get_values(index: Sequence[int]) -> List[float]:
+        return [vals[index[0]], vals[-1 - index[1]]]
+
+    # Generate the set of all possible indices.
+    all_indices = set((i, j) for i in range(len(vals)) for j in range(len(vals)))
+    for (i1, j1) in all_indices:
+        # Generate the set of all indices that dominates the current index.
+        dominating_indices = set((i2, j2) for i2 in range(i1 + 1) for j2 in range(j1 + 1))
+        dominating_indices -= {(i1, j1)}
+
+        for (i2, j2) in dominating_indices:
+            t1 = _create_trial(get_values([i1, j1]))
+            t2 = _create_trial(get_values([i2, j2]))
+            assert _dominates(t2, t1, directions)
+
+        for (i2, j2) in all_indices - dominating_indices:
+            t1 = _create_trial(get_values([i1, j1]))
+            t2 = _create_trial(get_values([i2, j2]))
+            assert not _dominates(t2, t1, directions)
```

Comment:
It is difficult for me to follow the logic that determine which points dominate `(i1, j1)`. I guess `get_values` increases the complexity. How about removing it, for example, as follows?

```suggestion
    # Generate the set of all possible indices.
    all_indices = set((i, j) for i in range(len(vals)) for j in range(len(vals)))
    for (i1, j1) in all_indices:
        # Generate the set of all indices that dominates the current index.
        dominating_indices = set((i2, j2) for i2 in range(i1 + 1) for j2 in range(j1, len(vals)))
        dominating_indices -= {(i1, j1)}

        for (i2, j2) in dominating_indices:
            t1 = _create_trial((vals[i1], vals[j1]))
            t2 = _create_trial((vals[i2], vals[j2]))
            assert _dominates(t2, t1, directions)

        for (i2, j2) in all_indices - dominating_indices:
            t1 = _create_trial((vals[i1], vals[j1]))
            t2 = _create_trial((vals[i2], vals[j2]))
            assert not _dominates(t2, t1, directions)
```

In addition, I couldn't understand the relationship between the indices and values before drawing the following table. IMO, it would be helpful for developers to understand the logic if we explain the logic with a table.

- `directions = ["minimize", "maximize"]`
- ✔ dominates 🎯, which corresponds to `dominating_indices = set((i2, j2) for i2 in range(i1 + 1) for j2 in range(j1, len(vals)))` where `i1=-1, j1=1`.

| `vals[i]` \ `vals[j]`  | - `inf`  | -1 | 1  | `inf` |
| --- | --- | --- | --- | --- |
| - `inf` |  |  | ✔ | ✔ |
| -1 | | | 🎯 | ✔ |
| 1 | | | | |
| `inf` | | | | |

Reply:
Thanks!